src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveNotificationListener.cs

```diff
@@ -56,190 +58,141 @@ internal class FileMoveNotificationListener : IFileMoveNotificationListener
+                        }
                     }
-
-                    actions.Add((documentAction, filename));
                 }
-
-                return actions;
             }
         }
 
         public async Task OnAfterFileMoveAsync()
         {
-            if (_actions is { Count: not 0 } && await CheckUserConfirmationAsync())
+            if (!_renameActionSets.Any() || !await IsEnabledOrConfirmedAsync())
             {
-                ApplyNamespaceUpdateActions();
+                return;
             }
 
-            return;
-
-            async Task<bool> CheckUserConfirmationAsync()
+            _ = _threadingService.JoinableTaskFactory.RunAsync(async () =>
             {
-                ISettingsManager settings = await _settingsManagerService.GetValueAsync();
+                // The wait service requires the main thread to run.
+                await _threadingService.SwitchToUIThread();
+                // Displays a dialog showing the progress of updating the namespaces in the files.
+                _waitService.Run(
+                    title: string.Empty,
+                    message: _renameMessage!,
+                    allowCancel: true,
+                    asyncMethod: ApplyRenamesAsync,
+                    totalSteps: _renameActionSets.Count);
+            });
 
-                bool promptNamespaceUpdate = settings.GetValueOrDefault(VsToolsOptions.OptionPromptNamespaceUpdate, true);
-                bool enabledNamespaceUpdate = settings.GetValueOrDefault(VsToolsOptions.OptionEnableNamespaceUpdate, true);
+            return;
 
-                if (!enabledNamespaceUpdate || !promptNamespaceUpdate)
+            async Task ApplyRenamesAsync(IWaitContext context)
+            {
+                CancellationToken token = context.CancellationToken;
+                await TaskScheduler.Default;
+                // WORKAROUND: We don't yet have a way to wait for the changes to propagate to Roslyn, tracked by https://github.com/dotnet/project-system/issues/3425
+                // Instead, we wait for the IntelliSense stage to finish for the entire solution.
+                IVsOperationProgressStatusService statusService = await _operationProgressService.GetValueAsync(token);
+                await statusService.GetStageStatus(CommonOperationProgressStageIds.Intellisense).WaitForCompletionAsync().WithCancellation(token);
+                // After waiting, a "new" published Solution is available.
+                Solution solution = _workspace.CurrentSolution;
+
+                int currentStep = 1;
+                foreach (KeyValuePair<string, RenameDocumentActionSet> renameActionSet in _renameActionSets)
                 {
-                    return enabledNamespaceUpdate;
-                }
+                    // Display the filename being updated to the user in the progress dialog.
+                    context.Update(currentStep: currentStep++, progressText: Path.GetFileName(renameActionSet.Key));
```

Comment:
You mentioned in a comment above that the key is the file name, yet here you're pulling the file name?

Reply:
The key is the full file path on disk of the source of the item being moved. I've updated comments to reflect this (that I'll commit after testing the other changes I've made).